### PR TITLE
prevent withdrawal with no rewards

### PR DESCRIPTION
**Pull request by wolverineks**

### CODE REVIEW

**Review by stackchain — APPROVED** [review]

**Review by banklesss — APPROVED** [review]

File: src/yoroi-wallets/walletManager.ts [hunk 1]
```diff
@@ -391,35 +390,6 @@ export class WalletManager {
 
     return this.saveWallet(id, name, wallet, networkId, implementationId)
   }
-
```
Comment (wolverineks) [inline src/yoroi-wallets/walletManager.ts:394]: cleanups

File: storybook/mocks/storage.ts [hunk 2]
```diff
@@ -3,7 +3,7 @@ import storage from '../../src/legacy/storage'
 
 export const mockEncryptedStorage: WalletEncryptedStorage = {
   rootKey: {
-    read: (password) => password === 'pasword' ? Promise.resolve('') : Promise.reject(new Error('Invalid Password')),
```
Comment (wolverineks) [inline storybook/mocks/storage.ts:6]: fix typo

File: src/yoroi-wallets/cardano/types.ts [hunk 3]
```diff
@@ -9,12 +9,12 @@ import {
   AccountStates,
   StakePoolInfoRequest,
   StakePoolInfosAndHistories,
+  StakingInfo,
   StakingStatus,
   YoroiSignedTx,
   YoroiUnsignedTx,
 } from '../types'
 import type {
-  AddressedUtxo,
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/types.ts:17]: cleanup

File: src/yoroi-wallets/cardano/shelley/delegationUtils.ts [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
 
+import {CardanoAddressedUtxo} from '@emurgo/yoroi-lib'
 import {sortBy} from 'lodash'
 
 import assert from '../../../legacy/assert'
 import {ObjectValues} from '../../../legacy/flow'
 import {Logger} from '../../../legacy/logging'
 import {normalizeToAddress} from '../../../legacy/utils'
 import {StakingStatus} from '../../types'
-import type {AddressedUtxo} from '../../types/other'
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/shelley/delegationUtils.ts:10]: cleanup

File: src/yoroi-wallets/cardano/shelley/delegationUtils.ts [hunk 5]
```diff
@@ -41,10 +41,10 @@ const addrContainsAccountKey = async (
 
 export const filterAddressesByStakingKey = async (
   stakingKey: CardanoTypes.StakeCredential,
-  utxos: ReadonlyArray<AddressedUtxo>,
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/shelley/delegationUtils.ts:44]: cleanup

File: src/yoroi-wallets/cardano/shelley/delegationUtils.ts [hunk 6]
```diff
@@ -41,10 +41,10 @@ const addrContainsAccountKey = async (
...
+  utxos: ReadonlyArray<CardanoAddressedUtxo>,
   acceptTypeMismatch: boolean,
 ) => {
-  const result: Array<AddressedUtxo> = []
+  const result: Array<CardanoAddressedUtxo> = []
 
   for (const utxo of utxos) {
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/shelley/delegationUtils.ts:49]: cleanup

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 7]
```diff
@@ -491,27 +491,38 @@ export class ShelleyWallet extends Wallet implements WalletInterface {
     return addressedUtxos
   }
 
-  async getLegacyAddressedUtxos() {
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:494]: cleanup

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 8]
```diff
@@ -433,7 +433,7 @@ export class ShelleyWallet extends Wallet implements WalletInterface {
   async getAllUtxosForKey() {
     return filterAddressesByStakingKey(
       await CardanoMobile.StakeCredential.fromKeyhash(await (await this.getStakingKey()).hash()),
-      await this.getLegacyAddressedUtxos(),
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:436]: cleanup

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 9]
```diff
@@ -491,27 +491,38 @@ export class ShelleyWallet extends Wallet implements WalletInterface {
...
-  }
-
   getDelegationStatus() {
     if (!this.transactionCache) throw new Error('invalid wallet state')
     if (this.rewardAddressHex == null) throw new Error('reward address is null')
     const certsForKey = this.transactionCache.perRewardAddressCertificates[this.rewardAddressHex]
     return Promise.resolve(getDelegationStatus(this.rewardAddressHex, certsForKey))
   }
 
+  async getStakingInfo(): Promise<StakingInfo> {
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:501]: promote getStakingInfo from utility function to wallet method
moved from `StakePoolInfos.tsx`

File: src/Dashboard/WithdrawStakingRewards/WithdrawStakingRewards.tsx [hunk 10]
```diff
@@ -47,16 +49,25 @@ export const WithdrawalTxForm = ({
...
   const strings = useStrings()
-  const [deregister, setDeregister] = React.useState(false)
-  const {isLoading} = useWithdrawalTx({wallet, deregister}, {onSuccess: (withdrawalTx) => onDone(withdrawalTx)})
+  const [deregister, setDeregister] = React.useState<boolean>()
+  const {isLoading} = useWithdrawalTx(
+    {wallet, deregister},
+    {enabled: deregister != null, onSuccess: (withdrawalTx) => onDone(withdrawalTx)},
+  )
+
+  const hasRewards =
```
Comment (wolverineks) [inline src/Dashboard/WithdrawStakingRewards/WithdrawStakingRewards.tsx:60]: this is the fix 1/3

File: src/Dashboard/WithdrawStakingRewards/WithdrawStakingRewards.tsx [hunk 11]
```diff
@@ -47,16 +49,25 @@ export const WithdrawalTxForm = ({
...
+      ? Quantities.isGreaterThan(stakingInfo.rewards, '0')
+      : false
 
   return (
     <DangerousAction
       title={strings.warningModalTitle}
       alertBox={{content: [strings.warning1, strings.warning2, strings.warning3]}}
       primaryButton={{
-        disabled: isLoading,
+        disabled: !hasRewards || isLoading,
```
Comment (wolverineks) [inline src/Dashboard/WithdrawStakingRewards/WithdrawStakingRewards.tsx:70]: this is the fix 2/3

File: src/components/DangerousActionModal/DangerousActionModal.tsx [hunk 12]
```diff
@@ -75,6 +75,7 @@ export const DangerousAction = ({
           block
           onPress={primaryButton.onPress}
           title={primaryButton.label}
+          disabled={primaryButton.disabled}
```
Comment (wolverineks) [inline src/components/DangerousActionModal/DangerousActionModal.tsx:78]: this is the fix, 3/3

File: src/Dashboard/StakePoolInfos.tsx [hunk 13]
```diff
@@ -31,68 +32,34 @@ const styles = StyleSheet.create({
...
-  status: 'registered'
-}
-type Staked = {
-  status: 'staked'
-  poolId: string
-  amount: string
-  rewards: string
-}
-
-const getStakePoolInfo = async (wallet: YoroiWallet) => {
```
Comment (wolverineks) [inline src/Dashboard/StakePoolInfos.tsx:48]: promoted to wallet method

File: src/Dashboard/StakePoolInfos.tsx [hunk 14]
```diff
@@ -31,68 +32,34 @@ const styles = StyleSheet.create({
   },
 })
 
-type StakingInfo = Registered | Staked | NotRegistered
```
Comment (wolverineks) [inline src/Dashboard/StakePoolInfos.tsx:34]: moved to types

File: src/Dashboard/StakePoolInfos.tsx [hunk 15]
```diff
@@ -129,7 +94,10 @@ export const useStakingTx = (
       if (!accountState) throw new Error('Account state not found')
 
       const stakingUtxos = await wallet.getAllUtxosForKey()
-      const amountToDelegate = sum([...stakingUtxos.map((utxo) => utxo.amount), accountState.remainingAmount])
+      const amountToDelegate = Quantities.sum([
```
Comment (wolverineks) [inline src/Dashboard/StakePoolInfos.tsx:97]: cleanups

File: src/Dashboard/StakePoolInfos.tsx [hunk 16]
```diff
@@ -103,7 +70,7 @@ export const useStakingInfo = (
 
 export const useStakePoolIds = (
   wallet: YoroiWallet,
-  options?: UseQueryOptions<StakingInfo, Error, StakingInfo, [string, string]>,
+  options?: UseQueryOptions<StakingInfo, Error, StakingInfo, [string, 'stakingInfo']>,
```
Comment (wolverineks) [inline src/Dashboard/StakePoolInfos.tsx:73]: cleanups

**Review by stackchain — APPROVED** [review]

**Pull request merged**